include/merlin/core_kernels/lookup.cuh

```diff
@@ -1017,24 +1163,53 @@ __global__ void lookup_kernel(const Table<K, V, S>* __restrict table,
     occupy_result = find_without_lock<K, V, S, TILE_SIZE>(
         g, bucket, find_key, start_idx, key_pos, src_lane, bucket_max_size);
 
-    if (occupy_result == OccupyResult::DUPLICATE) {
+    bool found = (occupy_result == OccupyResult::DUPLICATE);
+    if (found) {
       if (rank == src_lane) {
         *(values + key_idx) = (bucket->vectors + key_pos * dim);
         if (scores != nullptr) {
           *(scores + key_idx) =
               bucket->scores(key_pos)->load(cuda::std::memory_order_relaxed);
         }
-        if (found != nullptr) {
-          *(found + key_idx) = true;
-        }
       }
     } else {
       if (rank == 0) {
         *(values + key_idx) = nullptr;
       }
     }
+
+    if (rank == 0) {
+      found_functor(key_idx, find_key, found);
+    }
   }
 }
 
+template <class K, class V, class S, uint32_t TILE_SIZE = 4>
+__global__ void lookup_kernel(const Table<K, V, S>* __restrict table,
+                              Bucket<K, V, S>* buckets,
+                              const size_t bucket_max_size,
+                              const size_t buckets_num, const size_t dim,
+                              const K* __restrict keys, V** __restrict values,
+                              S* __restrict scores, bool* __restrict founds,
+                              int* __restrict dst_offset, size_t N) {
+  FoundFunctorV1<K> found_functor(founds);
+  lookup_kernel_impl<K, V, S, decltype(found_functor), TILE_SIZE>(
+      table, buckets, bucket_max_size, buckets_num, dim, keys, values, scores,
+      found_functor, dst_offset, N);
+}
+
+template <class K, class V, class S, uint32_t TILE_SIZE = 4>
+__global__ void lookup_kernel(
+    const Table<K, V, S>* __restrict table, Bucket<K, V, S>* buckets,
+    const size_t bucket_max_size, const size_t buckets_num, const size_t dim,
+    const K* __restrict keys, V** __restrict values, S* __restrict scores,
+    K* __restrict missed_keys, int* __restrict missed_indices,
+    int* __restrict missed_size, int* __restrict dst_offset, size_t N) {
+  FoundFunctorV2<K> found_functor(missed_keys, missed_indices, missed_size);
```

Comment:
atomicadd in FoundFunctorV2 runs on all warps when the key is not found, which can cause performance degradation when there are lots of missed key. Isn't it better to split the missed_keys size used by found_functor from n to the bucket size, and then merge according to missed_size at last kernel finishing?

Reply:
Good point, I'll work on refining this code.

Reply:
> Good point, I'll work on refining this code.

What was the result? Does this optimization show up in performance tests?

Reply:
I apologize for the late reply, but I have verified your suggestion. Due to the additional calculations brought by the bucket separation, the benchmark results did not show performance improvement.